pkg/prometheus/resource_selector.go

```diff
@@ -120,65 +120,79 @@ func (rs *ResourceSelector) SelectServiceMonitors(ctx context.Context, listFn Li
 	res := make(map[string]*monitoringv1.ServiceMonitor, len(serviceMonitors))
 	for namespaceAndName, sm := range serviceMonitors {
 		var err error
+		rejectFn := func(serviceMonitor *monitoringv1.ServiceMonitor, err error) {
+			rejected++
+			level.Warn(rs.l).Log(
+				"msg", "skipping servicemonitor",
+				"error", err.Error(),
+				"servicemonitor", namespaceAndName,
+				"namespace", objMeta.GetNamespace(),
+				"prometheus", objMeta.GetName(),
+			)
+			rs.eventRecorder.Eventf(sm, v1.EventTypeWarning, operator.InvalidConfigurationEvent, "ServiceMonitor %s was rejected due to invalid configuration: %v", sm.GetName(), err)
+		}
 
 		for i, endpoint := range sm.Spec.Endpoints {
 			// If denied by Prometheus spec, filter out all service monitors that access
 			// the file system.
 			if cpf.ArbitraryFSAccessThroughSMs.Deny {
 				if err = testForArbitraryFSAccess(endpoint); err != nil {
-					break
+					rejectFn(sm, err)
+					continue
```

Comment:
IIUC it introduces a bug: if the service monitor contains 1 invalid endpoint followed by 1 valid endpoint, it will still be selected.

Reply:
Thanks @simonpasquier https://github.com/prometheus-operator/prometheus-operator/pull/6339